cpp/src_prims/stats/histogram.h

```diff
@@ -0,0 +1,488 @@
+      gmemHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                              stream);
+      break;
+    case HistTypeSmemWarp:
+      smemWarpHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                                  stream);
+      break;
+    case HistTypeSmem:
+      smemHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                              stream);
+      break;
+    case HistTypeSmemBits16:
+      smemBitsHist<DataT, BinnerOp, IdxT, 16, VecLen>(bins, nbins, data, n, op,
+                                                      tpb, stream);
+      break;
+    case HistTypeSmemBits8:
+      smemBitsHist<DataT, BinnerOp, IdxT, 8, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemBits4:
+      smemBitsHist<DataT, BinnerOp, IdxT, 4, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemBits2:
+      smemBitsHist<DataT, BinnerOp, IdxT, 2, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemBits1:
+      smemBitsHist<DataT, BinnerOp, IdxT, 1, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemHash:
+      smemHashHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                                  stream);
+      break;
+    default:
+      ASSERT(false, "histogram: Invalid type passed '%d'!", type);
+  };
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT>
+void histogramImpl(HistType type, int* bins, IdxT nbins, const DataT* data,
+                   IdxT n, int tpb, cudaStream_t stream, BinnerOp op) {
+  size_t bytes = n * sizeof(DataT);
+  if (n <= 0) return;
+  if (16 / sizeof(DataT) && bytes % 16 == 0) {
+    histogramVecLen<DataT, BinnerOp, IdxT, 16 / sizeof(DataT)>(
+      type, bins, nbins, data, n, tpb, stream, op);
+  } else if (8 / sizeof(DataT) && bytes % 8 == 0) {
+    histogramVecLen<DataT, BinnerOp, IdxT, 8 / sizeof(DataT)>(
+      type, bins, nbins, data, n, tpb, stream, op);
+  } else if (4 / sizeof(DataT) && bytes % 4 == 0) {
+    histogramVecLen<DataT, BinnerOp, IdxT, 4 / sizeof(DataT)>(
+      type, bins, nbins, data, n, tpb, stream, op);
+  } else if (2 / sizeof(DataT) && bytes % 2 == 0) {
+    histogramVecLen<DataT, BinnerOp, IdxT, 2 / sizeof(DataT)>(
+      type, bins, nbins, data, n, tpb, stream, op);
+  } else {
+    histogramVecLen<DataT, BinnerOp, IdxT, 1>(type, bins, nbins, data, n, tpb,
+                                              stream, op);
+  }
+}
+
+template <typename IdxT>
+HistType selectBestHistAlgo(IdxT nbins) {
+  size_t smem = maxSharedMem();
+  size_t requiredSize = nbins * sizeof(unsigned);
```

Comment:
This whole function should definitely be simplified.

Reply:
if you have any suggestions, I'm all ears.

Reply:
At the very least, the cascade of ifs can be replaced by a loop, with the return value being selected from an array. It may also be possible to compute the index into that array directly.

Reply:
Done.